2.TCGA-MLexample-covariates.py

```diff
@@ -0,0 +1,350 @@
+    """
+    scores = mad(x) 
+    return scores, np.array([np.NaN]*len(scores))
+
+
+# ## Define pipeline and Cross validation model fitting
+
+# In[192]:
+
+# Parameter Sweep for Hyperparameters
+param_grid = {
+    'classify__loss': ['log'],
+    'classify__penalty': ['elasticnet'],
+    'classify__alpha': [10 ** x for x in range(-3, 1)],
+    'classify__l1_ratio': [0, 0.2, 0.8, 1],
+}
+
+pipeline = Pipeline(steps=[
+    ('standardize', StandardScaler()),
+    ('classify', SGDClassifier(random_state=0, class_weight='balanced'))
+])
+
+models = ['full', 'expressions', 'covariates']
+
+cv_pipelines = {mod: GridSearchCV(estimator=pipeline, 
+                             param_grid=param_grid, 
+                             n_jobs=1, 
+                             scoring='roc_auc') for mod in models}
+
+
+# In[193]:
+
+get_ipython().run_cell_magic('time', '', "for model, pipeline in cv_pipelines.items():\n    print('Fitting CV for model: {0}'.format(model))\n    pipeline.fit(X=X_partitions.get(model).get('train'), y=y_train)\n# cv_pipeline_full.fit(X=X_train_full, y=y_train)")
+
+
+# In[194]:
+
+# Best Params
+for model, pipeline in cv_pipelines.items():
+    print('{0}: {1:.3%}'.format(model, pipeline.best_score_))
+
+    # Best Params
+    print(pipeline.best_params_)
+
+
+# ## Visualize hyperparameters performance
+
+# In[195]:
+
+cv_results_df_dict = {model: 
+    pd.concat([
+        pd.DataFrame(pipeline.cv_results_),
+        pd.DataFrame.from_records(pipeline.cv_results_['params']),
+    ], axis='columns') for model, pipeline in cv_pipelines.items()}
+
+model = 'full'
+
+cv_results_df_dict[model].head(2)
+
+
+# In[196]:
+
+# Cross-validated performance heatmap
+model = 'full'
+
+cv_score_mat = pd.pivot_table(cv_results_df_dict[model],
+                              values='mean_test_score', 
+                              index='classify__l1_ratio',
+                              columns='classify__alpha')
+ax = sns.heatmap(cv_score_mat, annot=True, fmt='.1%')
+ax.set_xlabel('Regularization strength multiplier (alpha)')
+ax.set_ylabel('Elastic net mixing parameter (l1_ratio)');
+
+
+# ## Use Optimal Hyperparameters to Output ROC Curve
+
+# In[197]:
+
+y_pred_dict = {
+    model: {
+        'train': pipeline.decision_function(X_partitions[model]['train']),
+        'test':  pipeline.decision_function(X_partitions[model]['test'])
+    } for model, pipeline in cv_pipelines.items()
+}
+
+def get_threshold_metrics(y_true, y_pred):
+    roc_columns = ['fpr', 'tpr', 'threshold']
+    roc_items = zip(roc_columns, roc_curve(y_true, y_pred))
+    roc_df = pd.DataFrame.from_items(roc_items)
+    auroc = roc_auc_score(y_true, y_pred)
+    return {'auroc': auroc, 'roc_df': roc_df}
+
+metrics_dict = {    
+    model: {
+        'train': get_threshold_metrics(y_train, y_pred_dict[model]['train']),
+        'test':  get_threshold_metrics(y_test, y_pred_dict[model]['test'])
+    } for model in y_pred_dict.keys()
+}
+
+
+# In[198]:
+
+# Plot ROC
```

Comment:
Coloring by model and linetype by test/train would be awesome here. Not sure if this is easy to add.

We may also want to consider replacing this plot with your vega-lite viz. Or maybe even [bokeh](http://bokeh.pydata.org/). Lot's of [options](https://www.youtube.com/watch?v=FytuB8nFHPQ), not straightforward which is best.